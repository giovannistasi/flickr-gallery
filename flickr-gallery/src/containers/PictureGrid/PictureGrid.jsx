import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types'
import styled from 'styled-components';
import throttle from 'lodash.throttle';

import { API } from '../../store/middlewares/apiService'

import Picture from '../../components/Picture'
import SelectedPicture from '../../components/SelectedPicture'
import PictureLoader from '../../components/PictureLoader';

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(400px, 1fr));
  row-gap: 5vmin;
  column-gap:5vim;
  justify-items: center;
  width:100%;
  margin: 5vmin 0;
  `

class PictureGrid extends Component {
  static propTypes = {
    pictures: PropTypes.array,
    selectedPicture: PropTypes.object,
    fetchPictures: PropTypes.func.isRequired,
    selectPicture: PropTypes.func.isRequired,
  };

  pageNumber;

  componentWillMount () {
    // Initial two calls to server
    this.props.fetchPictures()
    this.props.fetchPictures()
    //throttle method to avoid unnecesary calls to server
    this.lazyFetchPictures = throttle(this.props.fetchPictures, 500)

    this.listenForScrollAndFetch() // From this point on, only fetch on scroll
  }

  // fetch new pictures when scrolled almost till the bottom
  listenForScrollAndFetch = () => {
    window.addEventListener('scroll', (e) => {
      const isLoading = this.props.loading;
      const scrollAfterTreshold = window.innerHeight + window.scrollY >=
        document.documentElement.scrollHeight - window.innerHeight / 0.5;
      if (!isLoading && scrollAfterTreshold) {
        if (this.props.selectedTag) {
          // throttle(() => {
          this.props.fetchPicturesFromTag(this.props.selectedTag, this.props.pageNum)
          this.props.increasePageNum();
          // }, 500
          // )
        } else {
          console.log(this.props.selectedTag);
          this.lazyFetchPictures();
        }
      }
    })
  }

  renderPictures = (pictures) => {

    const pictureLoaders = 5 // Number of loaders (blank images) at the end of the grid
    return [
      ...pictures.map(p => <Picture key={p.id} picture={p} handleClick={this.selectPicture} />),
      this.renderPictureLoaders(pictureLoaders)
    ]
  }

  renderPictureLoaders = (num) => {
    // Calculate the opacity for each loader so it results in a nice scaling down grayscale
    let res = []
    const step = 1 / num;
    for (let i = 1; i <= num; i++) {
      let opacity = step * i
      if (opacity === 1) opacity -= 0.1 // We don't want opacity === 1, too bright!
      res.push(opacity)
    }
    return res.reverse().map(n => <PictureLoader opacity={n.toFixed(1)} key={n} />)
  }

  selectPicture = (picture) => {
    this.props.selectPicture(picture)
  }

  render () {
    const { pictures, selectedPicture } = this.props;


    return (
      <Container>
        {pictures && this.renderPictures(pictures)}
        {selectedPicture &&
          <SelectedPicture selectedPicture={selectedPicture}
            handleClick={this.selectPicture} />}
      </Container>
    )
  }
}

const mapStateToProps = (state) => {

  return {
    pictures: state.pictures.pictureList,
    selectedPicture: state.pictures.selectedPicture,
    selectedTag: state.pictures.selectedTag,
    pageNum: state.pictures.pageNum,
    loading: state.pictures.loading
  }
}

const mapDispatchToProps = dispatch => ({
  fetchPictures: () => dispatch({
    type: 'FETCH_PICTURES',
    [API]: {
      path: '/pictures'
    }
  }),
  selectPicture: (picture) => dispatch({
    type: 'SELECT_PICTURE',
    data: picture
  }),
  fetchPicturesFromTag: (tag, pageNum) => dispatch({
    type: 'FETCH_PICTURES_FROM_TAG',
    [API]: {
      path: '/pictures-from-tags/:' + tag + '/:' + pageNum,
    }
  }),
  increasePageNum: () => dispatch({
    type: 'INCREASE_PAGE_NUM'
  })
})

// const mapDispatchToProps = {
//   fetchPictures, selectPicture
// }

// function fetchPictures() {
//   return {
//     type: 'FETCH_PICTURES',
//     [API]: {
//       path: '/pictures'
//     }
//   }
// }

export default connect(mapStateToProps, mapDispatchToProps)(PictureGrid)