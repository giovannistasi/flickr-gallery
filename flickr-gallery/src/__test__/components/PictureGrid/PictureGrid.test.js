import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react'
import { createStore } from 'redux'
import { Provider } from 'react-redux';
import reducer from '../../../store/reducers/index.js'
import PictureGrid from '../../../containers/PictureGrid/PictureGrid'

const mockPicture = {
  title: 'ciao',
  id: 1,
  ownerInfo: {
    profile: {
      first_name: 'Bob',
      last_name: 'Stasi',
      id: '1234'
    }
  }
}
const mockPicture2 = {
  title: 'ciao2',
  id: 2,
  ownerInfo: {
    profile: {
      first_name: 'Bob',
      last_name: 'Stasi',
      id: '1234'
    }
  }
}


const pictures = [mockPicture, mockPicture2]
const INITIAL_STATE = { pictures: { pictureList: pictures } }
function renderWithRedux (
  ui,
  { store = createStore(reducer,
    INITIAL_STATE) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  }
}

describe('Picture on click', () => {
  it('loads selectedPicture on click', async () => {
    const { container } = renderWithRedux(
      <PictureGrid picture={mockPicture} />
    )
    const PictureElement = await waitForElement(() => container.querySelector('.Picture'));
    fireEvent.click(PictureElement)
    expect(container.querySelector('.SelectedPicture__outer')).toBeTruthy();
  });
});

describe('Loading pictures on scroll', () => {
  it('should only fetch on scroll', () => {
    // scrollContainer.addEventListener('scroll', () => { /* some callback */ });
    const { container } = renderWithRedux(
      <PictureGrid picture={mockPicture} />
    );
    const mockScrollFetch = jest.fn(container.listenForScrollAndFetch)
    console.log(mockScrollFetch);
    fireEvent.scroll(container, { target: { scrollY: 1000 } });
    expect(mockScrollFetch).toHaveBeenCalled();
  });
})
