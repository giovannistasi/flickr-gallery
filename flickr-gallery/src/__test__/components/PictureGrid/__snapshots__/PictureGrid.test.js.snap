// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picture Grid loading should render selected picture correctly 1`] = `
<div
  class="SelectedPicture__outer"
>
  <div
    class="SelectedPicture__inner-container"
    alt="Open picture on Flickr"
  >
    <h1>
      ciao
    </h1>
    <img
      src="https://picsum.photos/300/?random"
      alt="ciao"
    />
    <a
      href="https://www.flickr.com/photos/undefined"
      alt="Open user profile on Flickr"
    >
      <h2>
        By 
        Bob Stasi
      </h2>
    </a>
  </div>
</div>
`;
